### ISSUE TITLE
deepEqual returning `undefined`

### ISSUE THREAD

**Issue by scripter-co** [sinonjs/samsam#25]
As `samsam` is going to be released with a breaking change, it would be good to tidy up some of the instances where `deepEqual` can return undefined.

Having taken a look through the code, it seems there is one condition that is not handled correctly, causing it to return undefined, that's [here](https://github.com/sinonjs/samsam/blob/master/lib/samsam.js#L244).

For instance, this test will fail:

```js
it("should fail when comparing an object to null", () => {
    var checkDeep = samsam.deepEqual({}, null);
    assert.equals(checkDeep, true);
});
```

The problem is that `identical` is set to return `undefined`. This is the same as `isSet` but that's handled correctly by explicitly returning `false` [here](https://github.com/sinonjs/samsam/blob/master/lib/samsam.js#L276). We could simply fix the issue by conditionally returning `undefined` from within `deepEqual` if we get undefined back from `identical`, or we could refactor `identical` to return false. My suggestion would be the latter whilst also fixing `isSet`.

This would then allow us to use `samsam.deepEqual` within the sinon project, resolving this issue: https://github.com/sinonjs/sinon/issues/1709

**Comment by mroderick** [comment]
Thank you for taking the time to carefully report this issue. I find it has a good amount of detail 👍 

I agree with your analysis as well as your suggestion for improvements: `identical` and `isSet` should only ever return a `Boolean` value.

Would you like to contribute a pull request to resolve this issue?

**Comment by scripter-co** [comment]
Yep. I'll get a PR open. Would be good to get #22 finished and merged first so I can add additional tests using `mocha`.

**Comment by mroderick** [comment]
#22 is merged, you're good to go!

**Comment by scripter-co** [comment]
@mroderick on it.

**Comment by mroderick** [comment]
This has been fixed with #27
